### PR TITLE
NAS-123903 / NFS41 ACL support for NFS4 protocol

### PR DESCRIPTION
**Pull request by bmeagherix**
### Background
ZFS currently exposes native ZFS nfsv4 ACL type through the `system.nfs4_acl_xdr` xattr.

A previous PR was "[Implement native NFSv4 ACLs in NFS server](https://github.com/truenas/linux/pull/87)" which allowed the ZFS ACL to be exposed thru the existing NFS 4.0 ACL support in `nfsd`.  One limitation is that the NFSv4.0 ACL specification does not allow the `acl_flag` to be included.

NFSv4.1 added support for a slightly more capable variant of ACL, the DACL (and SACL) which also includes support for an `acl_flag` field.

Upstream linux includes client-side support for DACLs, but **not** server-side support.  Because no in-kernel filesystems have suitable ACL support, it seems very unlikely that changes to `nfsd` to support DACLs will be accepted upstream.

### Overview
This PR adds:
- server-side DACL support, based on the underlying ZFS `system.nfs4_acl_xdr` xattr.
- client-side support for `system.nfs4_acl_xdr` xattr, utilizing the existing client DACL support (for transport).

This will permit existing tools that operate on the `system.nfs4_acl_xdr` xattr (e.g. `nfs4xdr_setfacl`) to operate on a NFS v4.1 mounted client.

### Implementation
The PR has been split into 5 initial commits.

1. Add a `flag` to `struct nfs4_acl` and move `enum nfs4_acl_type` for reuse in `nfsd`.
This will permit `struct nfs4_acl` to also support DACLs (or SACL).  The `flag` will remain zero for ACL.
We will use the `nfs4_acl_type` enum at various call-sites to determine whether we're operating on an ACL or DACL.  (This pattern was already being used by `nfs` client in the upstream kernel; move the enum so we can do likewise in `nfsd`.)

2. Add DACL support to `nfsd` for NFSv4.1+
DACL support is modeled on the existing ACL support, and the existing nfs infrastructure for attr handling does most of the heavy lifting here.
Some functions (e.g. `nfsd4_get_nfs4_acl`) have their signatures widened to include an `acl_type` enum parameter.  We add `FATTR4_WORD1_DACL` to the `NFSD4_1_SUPPORTED_ATTRS_WORD1` and `NFSD_WRITEABLE_ATTRS_WORD1` bitmasks to allow plumbing DACL through the various existing routines, but note that this is tempered by `IS_NFSV4ACL` checks.
4. Move some (iX-authored) `nfs4_acl_xdr` handling functions from _nfsd_ to _nfs_common_ for reuse
These were authored during the above-mentioned NFSv4.0 [PR](https://github.com/truenas/linux/pull/87) & used in `nfsd`.  We now want to use them in _both_ the nfs server _and_ client.
5. Add `isdir` parameter to `generate_nfs41acl_buf` and populate acl flags
When generating the (ZFS) buffer we also want to be able to set the `ACL_IS_DIR` flag in the acl_flag.  Since this is **not** included in the NFSv4.1 DACL acl_flag bits, we do not send it over the wire.  However, we can regenerate it if the inode in question refers to a directory.  Hence the plumbing.
7. Add nfs client-side support for `system.nfs4_acl_xdr`
The new accessor functions for the `system.nfs4_acl_xdr` xattr are implemented using the existing client-side support for DACL for transport of the DACL to/from the `nfsd`.
We then need to perform a `struct nfs4_acl` to/from DACL XDR, implemented by `nfs4_aclbuf_decode` and `nfs4_encode_acl_to_buf`.   (BTW, even if it were accessible, the existing `nfsd` code to do the equivalent functionality is _deeply_ embedded into the server-side infrastructure & unsuitable for reuse here.)  Only OWNER@, GROUP@, EVERYONE@ and numeric uids or gids are supported in the ACEs, but this turns out to not be a restriction here, as that is how the `system.nfs4_acl_xdr` presents/transfers them anyway.
Finally, encoding/decoding is required to translate the `struct nfs4_acl` to/from the ZFS-style ACL using the functions moved to nfs_common.

### Limitations / caveats
The existing upstream nfs client contains a [nfs4_cached_acl](https://elixir.bootlin.com/linux/v6.1.55/A/ident/nfs4_cached_acl).  This can hold either the ACL or DACL (or SACL).  If, for example, the DACL has been cached by the client and then the DACL is changed on the server **outside the scope of NFS**, then the client will have no knowledge of this change and will retain the old value for a period of time. (Appears to be cached several seconds to 10s of seconds, based on activity.)  This is the case even for upstream POSIX ACL-based NFSv4.0 ACL (`system.nfs4_acl`), and before this PR.

This can be worked around by e.g. fetching the ACL and then fetching the DACL again to force refresh the cache.

### CODE REVIEW

File: fs/nfs/nfs4proc.c [hunk 1]
```diff
@@ -7774,6 +7778,474 @@ static bool nfs4_xattr_list_nfs4_sacl(struct dentry *dentry)
...
+
+/*
+ * Decode the buffer containing the raw XDR data into struct nfs4_acl.
+ *
+ * This is similar to nfsd4_decode_acl
+ */
+static int nfs4_aclbuf_decode(void *buf, size_t buflen, enum nfs4_acl_type type,
+			      struct nfs4_acl **pacl)
+{
+	// int error = 0;
```
Comment (umsaleem) [inline fs/nfs/nfs4proc.c]: Maybe this line can be removed.

File: fs/nfs/nfs4proc.c [hunk 2]
```diff
@@ -7774,6 +7778,474 @@ static bool nfs4_xattr_list_nfs4_sacl(struct dentry *dentry)
...
+	},
+};
+
+static int nfs4_acl_get_whotype(char *p, u32 len)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(s2t_map); i++) {
+		if (s2t_map[i].stringlen == len &&
+				(memcmp(s2t_map[i].string, p, len) == 0))
```
Comment (ixhamza) [inline fs/nfs/nfs4proc.c]: Extra tab.

File: fs/nfs/nfs4proc.c [hunk 3]
```diff
@@ -7774,6 +7778,474 @@ static bool nfs4_xattr_list_nfs4_sacl(struct dentry *dentry)
...
+	acl_bytes = nfs4_acl_bytes(count);
+	acl = kmalloc(acl_bytes, GFP_KERNEL);
+	if (acl == NULL)
+		return -ENOMEM;
+
+	acl->flag = acl_flag;
+	acl->naces = count;
+	for (ace = acl->aces; ace < acl->aces + count; ace++) {
+		status = nfs4_decode_nfsace4((void **)&p, ace);
+		if (status)
```
Comment (ixhamza) [inline fs/nfs/nfs4proc.c]: I think `acl` needs to be free here.

Comment (umsaleem) [reply fs/nfs/nfs4proc.c]: I think we are returning `acl` by passing in `pacl`, so maybe not.

Comment (umsaleem) [reply fs/nfs/nfs4proc.c]: Ah, we are returning because of error, I guess we should free here.

Comment (bmeagherix) [reply fs/nfs/nfs4proc.c]: Yes, good catch!  Thanks.

**Review by umsaleem — APPROVED** [review]

**Comment by bmeagherix** [comment]
No longer wish to update linux 6.1 with these changes

**Pull request closed without merging**